File: tests/src/commonTest/kotlin/primitives/PingPrimitive.kt
package primitives

import natives.test.MyEnum
import natives.test.pingBoolean
import natives.test.pingByte
import natives.test.pingChar
import natives.test.pingDouble
import natives.test.pingFloat
import natives.test.pingInt
import natives.test.pingLong
import natives.test.pingString
import natives.test.pingShort
import withLib
import kotlin.test.Test
import kotlin.test.assertEquals

class PingPrimitive {

    @Test
    fun pingChar() = withLib {
        assertEquals('a', pingChar('a'))
    }

    @Test
    fun pingBoolean() = withLib {
        assertEquals(true, pingBoolean(true))
    }

    @Test
    fun pingByte() = withLib {
        assertEquals(99.toByte(), pingByte(99.toByte()))
    }

    @Test
    fun pingShort() = withLib {
        assertEquals(99.toShort(), pingShort(99.toShort()))
    }

    @Test
    fun pingInt() = withLib {
        assertEquals(99, pingInt(99))
    }

    @Test
    fun pingLong() = withLib {
        assertEquals(9223372036854775805L, pingLong(9223372036854775805L))
    }

    @Test
    fun pingFloat() = withLib {
        assertEquals(99f, pingFloat(99f))
    }

    @Test
    fun pingDouble() = withLib {
        assertEquals(99.0, pingDouble(99.0))
    }

    @Test
    fun pingString() = withLib {
        assertEquals("test string", pingString("test string"))
    }

    @Test
    fun pingEnum() = withLib {
        assertEquals(MyEnum.CASE2, natives.test.pingEnum(MyEnum.CASE2))
    }

}

File: modules/tests/natives/testcpp/src/primitives/ping_primitive.cpp
#include <api.hpp>

uint16_t ping_char(uint16_t arg) {
    return arg;
}

bool ping_boolean(bool arg) {
    return arg;
}

int8_t ping_byte(int8_t arg) {
    return arg;
}

uint8_t ping_ubyte(uint8_t arg) {
    return arg;
}

int16_t ping_short(int16_t arg) {
    return arg;
}

uint16_t ping_ushort(uint16_t arg) {
    return arg;
}

int32_t ping_int(int32_t arg) {
    return arg;
}

uint32_t ping_uint(uint32_t arg) {
    return arg;
}

int64_t ping_long(int64_t arg) {
    return arg;
}

uint64_t ping_ulong(uint64_t arg) {
    return arg;
}

float ping_float(float arg) {
    return arg;
}

double ping_double(double arg) {
    return arg;
}

KString ping_string(KString arg) {
    return arg;
}

KOptional<KString> ping_string_n(KOptional<KString> arg) {
    return arg;
}

MyEnum ping_enum(MyEnum arg) {
    return arg;
}

MyDictionary ping_dictionary(MyDictionary arg) {
    return arg;
}

KOptional<MyDictionary> ping_dictionary_n(KOptional<MyDictionary> arg) {
    return arg;
}

std::shared_ptr<IMyInterface> ping_interface(std::shared_ptr<IMyInterface> arg) {
    return arg;
}

KOptional<std::shared_ptr<IMyInterface>> ping_interface_n(KOptional<std::shared_ptr<IMyInterface>> arg) {
    return arg;
}
